<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="igvR">
<title>Use a Custom Genome • igvR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Use a Custom Genome">
<meta property="og:description" content="igvR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.17.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/alzheimersVariantsNearMEF2C.html">Explore VCF variants, GWAS snps, promoters and histone marks around the  MEF2C gene in Alzheimers Disease</a>
    <a class="dropdown-item" href="../articles/basicIntro.html">Introduction: a simple demo</a>
    <a class="dropdown-item" href="../articles/chooseStockOrCustomGenome.html">Choose aStock or Custom Genome</a>
    <a class="dropdown-item" href="../articles/ctcfChIP.html">Explore CTCF ChIP-seq alignments, MACS2 narrowPeaks, Motif Matching and H3K4me3 methylation</a>
    <a class="dropdown-item" href="../articles/ctcfChipSeq-old.html">Explore ChIP-seq alignments from a bam file, MACS2 narrowPeaks, conservation, H3K4me3 methylation and motif matching</a>
    <a class="dropdown-item" href="../articles/customGenome.html">Use a Custom Genome</a>
    <a class="dropdown-item" href="../articles/pairedEnd.html">Paired-end Interaction Tracks</a>
    <a class="dropdown-item" href="../articles/stockGenome.html">Use a Stock Genome</a>
    <a class="dropdown-item" href="../articles/ucscTableBrowser.html">Obtain and Display H3K4Me3 K562 track from UCSC table browser</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="customGenome_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Use a Custom Genome</h1>
                        <h4 data-toc-skip class="author">Paul Shannon</h4>
            
            <h4 data-toc-skip class="date">2022-07-24</h4>
      
      
      <div class="d-none name"><code>customGenome.Rmd</code></div>
    </div>

    
    
<style>
.main-container { width: 1000px; max-width:2800px;}
</style>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>With a properly configured web server, and a few annotation files, you can use igvR with any genome of interest. We demonstrate that here with covid-19, the SARS-CoV-2 genome, obtained from the <a href="https://www.ncbi.nlm.nih.gov/sars-cov-2" class="external-link">NCBI</a>, and hosted on an nginx webserver running at the Institute for Systems Biology.</p>
</div>
<div class="section level2">
<h2 id="explicit-loading-of-custom-hg38">Explicit loading of “custom” hg38<a class="anchor" aria-label="anchor" href="#explicit-loading-of-custom-hg38"></a>
</h2>
<p>We begin, however, by showing the explicit configuration required to loading the hg38 annotated genome hosted by igv.org, to familiarize you with the full range of parameters you can use if you have the data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://paul-shannon.github.io/igvR/" class="external-link">igvR</a></span><span class="op">)</span></span>
<span><span class="va">igv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/igvR-class.html">igvR</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://paul-shannon.github.io/BrowserViz/reference/setBrowserWindowTitle.html" class="external-link">setBrowserWindowTitle</a></span><span class="op">(</span><span class="va">igv</span>, <span class="st">"hg38 explicit"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setCustomGenome.html">setCustomGenome</a></span><span class="op">(</span><span class="va">igv</span>,</span>
<span>                id<span class="op">=</span><span class="st">"hg38"</span>,</span>
<span>                genomeName<span class="op">=</span><span class="st">"Human (GRCh38/hg38)"</span>,</span>
<span>                fastaURL<span class="op">=</span><span class="st">"https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg38/hg38.fa"</span>,</span>
<span>                fastaIndexURL<span class="op">=</span><span class="st">"https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg38/hg38.fa.fai"</span>,</span>
<span>                cytobandURL<span class="op">=</span><span class="st">"https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg38/cytoBandIdeo.txt"</span>,</span>
<span>                chromosomeAliasURL<span class="op">=</span><span class="cn">NA</span>,</span>
<span>                geneAnnotationName<span class="op">=</span><span class="st">"Refseq Genes"</span>,</span>
<span>                geneAnnotationURL<span class="op">=</span><span class="st">"https://s3.amazonaws.com/igv.org.genomes/hg38/refGene.txt.gz"</span>,</span>
<span>                geneAnnotationTrackHeight<span class="op">=</span><span class="fl">500</span>,</span>
<span>                geneAnnotationTrackColor<span class="op">=</span><span class="st">"darkBlue"</span>,</span>
<span>                initialLocus<span class="op">=</span><span class="st">"chr5:88,621,308-89,001,037"</span>,</span>
<span>                visibilityWindow<span class="op">=</span><span class="fl">5000000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="use-the-sars-cov-2-genome">Use the SARS-CoV-2 genome<a class="anchor" aria-label="anchor" href="#use-the-sars-cov-2-genome"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">base.url</span> <span class="op">&lt;-</span> <span class="st">"https://gladki.pl/igvR/testFiles/sarsGenome"</span></span>
<span><span class="va">fasta.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base.url</span>,<span class="st">"Sars_cov_2.ASM985889v3.dna.toplevel.fa"</span><span class="op">)</span></span>
<span><span class="va">fastaIndex.file</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base.url</span>, <span class="st">"Sars_cov_2.ASM985889v3.dna.toplevel.fa.fai"</span><span class="op">)</span></span>
<span><span class="va">annotation.file</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base.url</span>, <span class="st">"Sars_cov_2.ASM985889v3.101.gff3"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.sleep.html" class="external-link">Sys.sleep</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/setCustomGenome.html">setCustomGenome</a></span><span class="op">(</span><span class="va">igv</span>,</span>
<span>                id<span class="op">=</span><span class="st">"Sars_cov_2"</span>,</span>
<span>                genomeName<span class="op">=</span><span class="st">"Sars_cov_2.ASM985889v3"</span>,</span>
<span>                fastaURL<span class="op">=</span><span class="va">fasta.file</span>,</span>
<span>                fastaIndexURL<span class="op">=</span><span class="va">fastaIndex.file</span>,</span>
<span>                geneAnnotationURL<span class="op">=</span><span class="va">annotation.file</span>,</span>
<span>                geneAnnotationName<span class="op">=</span><span class="st">"ASM985889v3"</span>,</span>
<span>                geneAnnotationTrackHeight<span class="op">=</span><span class="fl">500</span>,</span>
<span>                geneAnnotationTrackColor<span class="op">=</span><span class="st">"darkBlue"</span>,</span>
<span>                visibilityWindow<span class="op">=</span><span class="fl">30000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="configure-and-run-an-nginx-webserver-with-cors-and-byte-range-support">Configure and run an nginx webserver, with CORS and Byte-Range support<a class="anchor" aria-label="anchor" href="#configure-and-run-an-nginx-webserver-with-cors-and-byte-range-support"></a>
</h2>
<p>If you work with a novel organism, or use a non-standard assembly, then the open source free <a href="https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-open-source/?_ga=2.179640484.1012398307.1658611594-1156607707.1658611594" class="external-link">nginx web server</a> is a good choice. I have had some success with a simple Python Flask server, but recently ran into errors which I could not fix, and so switched to (and now recommend) nginx.</p>
<div class="section level3">
<h3 id="cors-cross-origin-resource-sharing">CORS: cross-origin resource sharing<a class="anchor" aria-label="anchor" href="#cors-cross-origin-resource-sharing"></a>
</h3>
<p>Javascript interpreters running in compliant web browsers will (with some exceptions) not load code or data, within a running script, if that code or data comes from a web host other than the host from which the script came.</p>
<p>The CORS protocol eases this restriction - counter-intuitively, to my mind. If the cross-origin data has a header which announces CORS saftey, then Javascript will proceed. Thus nginx (or some alternative server you use) must be CORS enabled</p>
</div>
<div class="section level3">
<h3 id="byte-range-support">Byte-range support<a class="anchor" aria-label="anchor" href="#byte-range-support"></a>
</h3>
<p>Genome files are often very large and therefore only read into igv.js and manageable chunks. Your webserver must respond to those “chunk” requests, called “byte-range” support.</p>
</div>
<div class="section level3">
<h3 id="a-sample-nginx-configuration-">A sample nginx configuration.<a class="anchor" aria-label="anchor" href="#a-sample-nginx-configuration-"></a>
</h3>
<p>This is what we use at the Institute for Systems Biology to serve the SARS-CoV-2 genome and annotation. This file is called <em>default.conf</em> and is supplied in the Docker command shown below.</p>
<pre><code>server {
    listen       80;
    server_name  localhost;

    #access_log  /var/log/nginx/host.access.log  main;                                                                                                                                                              

    location / {
    root   /usr/share/nginx/html;
    index  index.html index.htm;
    # max requests per keepalive connection                                                                                                                                                                     
    keepalive_requests 55000;

    # awesome OPTIONS header                                                                                                                                                                                    
    if ($request_method = 'OPTIONS') {
        add_header "Access-Control-Allow-Origin" $http_origin;
        add_header "Vary" "Origin";
        add_header 'Access-Control-Allow-Credentials' 'true';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'DNT,Range,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Acopenept-Ranges,Content-Encoding,Content-R$
        # Tell client that this pre-flight info is valid for 20 days                                                                                                                                        
        add_header 'Access-Control-Max-Age' 1728000;
        add_header 'Content-Type' 'text/plain charset=UTF-8';
        add_header 'Content-Length' 0;
        return 204;
    }
        if ($request_method = 'POST') {
        add_header 'Access-Control-Allow-Origin' '*';
        add_header 'Access-Control-Allow-Credentials' 'true';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'DNT,Range,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Acopenept-Ranges,Content-Encoding,Content-R$
        add_header 'Access-Control-Expose-Headers' 'Accept-Ranges,Content-Encoding,Content-Length,Content-Range,Cache-Control,Content-Language,Content-Type,Expires,Last-Modified,Pragma,Date';
    }
        if ($request_method = 'GET') {
        add_header "Access-Control-Allow-Origin" $http_origin;
        add_header "Vary" "Origin";
        add_header 'Access-Control-Allow-Credentials' 'true';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'DNT,Range,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Acopenept-Ranges,Content-Encoding, Content-$
        add_header 'Access-Control-Expose-Headers' 'Accept-Ranges,Content-Encoding,Content-Length,Content-Range,Cache-Control,Content-Language,Content-Type,Expires,Last-Modified,Pragma,Date';
    }
    }

    # redirect server error pages to the static page /50x.html                                                                                                                                                      
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
    root   /usr/share/nginx/html;
    }
}</code></pre>
</div>
<div class="section level3">
<h3 id="run-nginx-out-of-a-docker-container-">Run nginx out of a docker container.<a class="anchor" aria-label="anchor" href="#run-nginx-out-of-a-docker-container-"></a>
</h3>
<p>The Docker container provided for free use by nginx.com is all you need. The <strong>-v</strong> options mount two host directories so that they are visible within the running container.</p>
<p>The virtual host setup, the DNS entry, and the proxying are not covered here.</p>
<pre><code>docker run -p 60050:80 \
      -v /yourConfigurationDirectory/fullPath/goesHere/default.conf:/etc/nginx/conf.d/default.conf:ro \
      -v /yourDataDirectory/fullPath/goesHere/:/usr/share/nginx/html:ro \
      --restart always \
      -d nginx</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] BiocStyle_2.25.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] knitr_1.39          magrittr_2.0.3      R6_2.5.1            ragg_1.2.2          rlang_1.0.4        </span></span>
<span><span class="co">#&gt;  [6] fastmap_1.1.0       stringr_1.4.0       tools_4.2.0         xfun_0.31           cli_3.3.0          </span></span>
<span><span class="co">#&gt; [11] jquerylib_0.1.4     systemfonts_1.0.4   htmltools_0.5.3     yaml_2.3.5          digest_0.6.29      </span></span>
<span><span class="co">#&gt; [16] rprojroot_2.0.3     pkgdown_2.0.6       bookdown_0.27       textshaping_0.3.6   BiocManager_1.30.18</span></span>
<span><span class="co">#&gt; [21] purrr_0.3.4         sass_0.4.2          fs_1.5.2            memoise_2.0.1       cachem_1.0.6       </span></span>
<span><span class="co">#&gt; [26] evaluate_0.15       rmarkdown_2.14      stringi_1.7.8       compiler_4.2.0      bslib_0.4.0        </span></span>
<span><span class="co">#&gt; [31] desc_1.4.1          jsonlite_1.8.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
